File: app/comum/telefone.py
"""Normalizacao de telefone brasileiro para uso em mensageria.

Forma canonica: somente digitos com prefixo 55 (sem '+').
Aceita nacional com DDD (10 ou 11 digitos) ou ja prefixado com 55 (12 ou 13).
"""


class TelefoneInvalido(ValueError):
    """Numero que nao pode ser usado como contato de mensageria no MVP."""


def normalizar(valor: str) -> str:
    digitos = "".join(c for c in valor if c.isdigit())
    if digitos.startswith("55") and len(digitos) in (12, 13):
        return digitos
    if len(digitos) in (10, 11):
        return "55" + digitos
    raise TelefoneInvalido(
        "Informe um telefone brasileiro com DDD "
        "(celular com 11 digitos ou fixo com 10)."
    )

File: frontend/src/painel/telefone.ts
export class TelefoneInvalido extends Error {
  constructor(mensagem = "Informe um telefone brasileiro com DDD (celular com 11 dígitos ou fixo com 10).") {
    super(mensagem);
    this.name = "TelefoneInvalido";
  }
}

export function normalizar(valor: string): string {
  const digitos = [...valor].filter((caractere) => caractere >= "0" && caractere <= "9").join("");
  if (digitos.startsWith("55") && (digitos.length === 12 || digitos.length === 13)) {
    return digitos;
  }
  if (digitos.length === 10 || digitos.length === 11) {
    return "55" + digitos;
  }
  throw new TelefoneInvalido();
}

export function telefoneUtilizavel(valor: string): boolean {
  try {
    normalizar(valor);
    return true;
  } catch {
    return false;
  }
}
